charmhelpers/contrib/openstack/deferred_events.py

```diff
@@ -0,0 +1,291 @@
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for managing deferred service events.
+
+This module is used to manage deferred service events from both charm actions
+and package actions.
+"""
+
+import collections
+import datetime
+import time
+
+import charmhelpers.core as ch_core
+import charmhelpers.contrib.openstack.policy_rcd as policy_rcd
+import charmhelpers.core.hookenv as hookenv
+import charmhelpers.core.host as host
+
+import subprocess
+
+
+ServiceEvent = collections.namedtuple(
+    'ServiceEvent',
+    'timestamp service reason action')
+
+
+class DeferredServiceEvents():
+    """A class to manage deferred service events."""
+
+    def __init__(self):
+        """Intialise deferred events class."""
+        self.events = self.load_events()
+
+
+class DeferredCharmServiceEvents(DeferredServiceEvents):
+    """Deferred service events from charm actions."""
+
+    KV_KEY = 'deferred_events'
+
+    def load_events(self):
+        """Load events previously requested by the charm.
+
+        :returns: List of events requested by the charm.
+        :rtype: List[ServiceEvent]
+        """
+        with ch_core.unitdata.HookData()() as t:
+            kv = t[0]
+            queued_restarts = kv.get(self.KV_KEY, [])
+        return [
+            ServiceEvent(
+                e['timestamp'], e['service'], e['reason'], e['action'])
+            for e in queued_restarts]
+
+    def add_event(self, service, action, event_reason=None):
+        """Record that a service event was requested but deferred.
+
+        :param services: Service which was to be restarted.
+        :type services: str
+        :param action: Action that was deferred
+        :type action: str
+        :param event_reason: Reason for needing to run the action.
+        :type event_reason: str
+        """
+        if not event_reason:
+            event_reason = 'Unknown'
+        timestamp = round(time.time())
+        self.events.append(ServiceEvent(
+            timestamp=timestamp,
+            service=service,
+            reason=event_reason,
+            action=action))
+        self.save_events()
+
+    def clear_deferred_events(self, services, action):
+        """Clear deferred events.
+
+        :param services: Services with deferred actions to clear.
+        :type services: List[str]
+        :param action: Action that was deferred
+        :type action: str
+        """
+        filtered_restarts = [
+            e
+            for e in self.events
+            if not (e.service in services and e.action == action)]
```

Comment:
I was being dense but I got caught up here on the "not". Could we add a bit more explanation in the docstring that we are removing the events for services AND keeping all other events.

Reply:
This code has been changed